### PR TITLE
Allow unary operation on non-constant variables

### PR DESCRIPTION
**Pull request by namannimmo10**
fixes one part of #355.
We also need to make this work at the backend.

### CODE REVIEW

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -2196,7 +2222,7 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
             throw SemanticAbort();
         }
         ASR::ttype_t *type = ASRUtils::TYPE(
-            ASR::make_Logical_t(al, x.base.base.loc, 4, nullptr, 0));
+            ASR::make_Logical_t(al, x.base.base.loc, 1, nullptr, 0));
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:2225]: Currently we use logical with kind=4 everywhere, so let's stick to that. If we are going to change it from 4 to 1, then let's do it as a separate PR and do it everywhere. I currently don't know the pros/cons.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 2]
```diff
@@ -1182,24 +1213,19 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
                     }
                     value = ASR::down_cast<ASR::expr_t>(
                         ASR::make_ComplexConstant_t(al, x.base.base.loc,
                         std::real(result), std::imag(result), operand_type));
                 }
             }
         }
-        tmp = ASR::make_UnaryOp_t(al, x.base.base.loc, op, operand, operand_type,
+        tmp = ASR::make_UnaryOp_t(al, x.base.base.loc, op, operand, result_type,
                               value);
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:1229]: I think that this will take `not 5` as an Integer operand and return a Logical. That is not the way to do it.

We have to treat UnaryOp just like we treat BinOp. It accepts and returns exactly the same type, always. And we use `Cast` to cast things correctly. So `not 5` should be represented as `UnaryOp(Cast(5, IntToLogical))`. 

Also, in our ASR refactoring, we split UnaryOp into `UnaryMinus` (accepts integer, float, complex and returns the same type) and `LogicalNot` (accepts and returns logical) and `IntegerBitNot` (I think that is the "Invert" thing). So that is how it should be implemented.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 3]
```diff
@@ -1102,13 +1102,50 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+            if (op == ASR::unaryopType::Not) {
+                result_type = logical_type;
+            } else if (op == ASR::unaryopType::Invert) {
+                throw SemanticError("Bad operand type for unary " +
+                    ASRUtils::unop_to_str(op) + ": " + ASRUtils::type_to_str_python(operand_type),
+                    x.base.base.loc);
+            } else {
+                result_type = operand_type;
+            }
+        }
```
Comment (czgdp1807) [inline src/lpython/semantics/python_ast_to_asr.cpp:1147]: Is it possible to merge `is_real` and `is_complex` in a single `if` branch? I see that their bodies are same.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 1 above)
Comment (czgdp1807) [reply src/lpython/semantics/python_ast_to_asr.cpp:2225]: Shifting from 4 to 1 will help in using lesser memory in the backend. That's one advantage. However, I agree that we should do this separately otherwise the diff gets confusing here.

File: tests/expr10.py [hunk 4]
```diff
@@ -7,6 +7,8 @@ def test_UnaryOp():
     b = not 5
     b = not -1
     b = not 0
+    b = True
+    a = -b
```
Comment (czgdp1807) [inline tests/expr10.py:11]: I think changes also include integer, real and complex types? Can we add tests for these types as well and try to cover as many kinds of each of these types as possible?

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 1 above)
Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp:2225]: See https://github.com/lcompilers/lpython/pull/476.

Comment (certik) [reply src/lpython/semantics/python_ast_to_asr.cpp:2225]: Note that how the backend actually implements `logical` with `kind=4` is a separate thing, and it can still implement it using just one byte --- in fact that's exactly what we do: https://github.com/lcompilers/lpython/blob/e5af88ae1454236105c416e08dcd9a0ae69b6f27/src/libasr/codegen/asr_to_llvm.cpp#L3375.

**Comment by namannimmo10** [comment]
Unary operation on logical types works in the current `main` branch. Thanks for your suggestions.
